scripts/tests/run_python_test.py

```diff
@@ -0,0 +1,160 @@
+
+
+def FindBinaryPath(name: str):
+    for path in pathlib.Path(DEFAULT_CHIP_ROOT).rglob(name):
+        if not path.is_file():
+            continue
+        if path.name != name:
+            continue
+        return str(path)
+
+    return None
+
+
+def EnqueueLogOutput(fp, tag, q):
+    for line in iter(fp.readline, b''):
+        timestamp = time.time()
+        if len(line) > len('[1646290606.901990]') and line[0:1] == b'[':
+            try:
+                timestamp = float(line[1:18].decode())
+                line = line[19:]
+            except Exception as ex:
+                pass
+        q.put((tag, line, datetime.datetime.fromtimestamp(
+            timestamp).isoformat(sep=" ")))
+    fp.close()
+
+
+def RedirectQueueThread(fp, tag, queue) -> threading.Thread:
+    log_queue_thread = threading.Thread(target=EnqueueLogOutput, args=(
+        fp, tag, queue))
+    log_queue_thread.start()
+    return log_queue_thread
+
+
+def DumpLogOutput(q: queue.Queue, timeout: int = 0):
+    deadline = time.time() + timeout
+    while True:
+        line = q.get((time.time() - deadline) > 0,
+                     max(deadline - time.time(), 0))
+        sys.stdout.buffer.write(
+            (f"[{line[2]}]").encode() + line[0] + line[1])
+        sys.stdout.flush()
+
+
+@click.command()
+@click.option("--app", type=str, default=None, help='Local application to use, omit to use external apps.')
+@click.option("--factoryreset", is_flag=True, help='Remove /tmp/chip* before running the tests.')
+@click.option("--wait-before-test", type=int, default=3, help='Time for the device to start before running the tests.')
+@click.option("--app-params", type=str, default='', help='The extra parameters passed to the device.')
+@click.option("--script", type=click.Path(exists=True), default=FindBinaryPath("mobile-device-test.py"), help='Test script to use.')
+@click.argument("script-args", nargs=-1, type=str)
+def main(app: str, factoryreset: bool, wait_before_test: int, app_params: str, script: str, script_args: typing.List[str]):
+    if factoryreset:
+        retcode = subprocess.call("rm -rf /tmp/chip*", shell=True)
+        if retcode != 0:
+            raise Exception("Failed to remove /tmp/chip* for factory reset.")
+
+    log_queue = queue.Queue()
+    log_cooking_threads = []
+
+    app_process = None
+    if app:
+        app = FindBinaryPath(app)
+        if app is None:
+            raise FileNotFoundError(f"{app} not found")
+        app_args = [app] + shlex.split(app_params)
+        logging.info(f"Execute: {app_args}")
+        app_process = subprocess.Popen(
+            app_args, stdout=subprocess.PIPE, stderr=subprocess.PIPE, bufsize=0)
+        log_cooking_threads.append(RedirectQueueThread(app_process.stdout,
+                                                       b"[\33[34mAPP \33[0m][\33[33mSTDOUT\33[0m]", log_queue))
+        log_cooking_threads.append(RedirectQueueThread(app_process.stderr,
```

Comment:
So, the issue with this approach is that the logs are actually not quite sorted in terms of timestamps:


```
[2022-03-15 10:34:36.700897][APP ][STDOUT][34912:34912] CHIP:EM: Flushed pending ack for MessageCounter:3392301 on exchange 1711i
[2022-03-15 10:34:37.696672][APP ][STDOUT][34912:34912] CHIP:DMG: Unblock report hold after min 1 seconds
[2022-03-15 10:34:39.708857][TEST][STDERR]INFO:PythonMatterControllerTEST:Received report from server: path: 1/6/0, value: False
[2022-03-15 10:34:39.710507][TEST][STDERR]INFO:PythonMatterControllerTEST:Testing on off cluster over resolved connection
[2022-03-15 10:34:39.710541][TEST][STDERR]INFO:PythonMatterControllerTEST:Sending On/Off commands to device 1 endpoint 1
[2022-03-15 10:34:39.718691][TEST][STDERR]INFO:PythonMatterControllerTEST:Testing non-controller APIs
[2022-03-15 10:34:39.718876][TEST][STDERR]INFO:PythonMatterControllerTEST:Test finished
[2022-03-15 10:34:39.704396][APP ][STDOUT][34912:34912] CHIP:EM: Received message of type 0x8 with protocolId (0, 1) and MessageCounter:3392302 on exchange 54832r
[2022-03-15 10:34:39.704442][APP ][STDOUT][34912:34912] CHIP:EM: Handling via exchange: 54832r, Delegate: 0x55d103cc3440
```

You can see above how the TEST side seems to indicate it has completed, BEFORE we see the actual app side logs.

This doesn't have to be fixed in this PR, but we should figure out how to fix it.

The timestamps of the logs are always stamped by the same clock, so you can use that as the sort critera.

You'd need some kind of temporary re-order buffer to allow you to buffer up those logs (perhaps for a second or so) before pumping out the re-ordered logs.